src/OrchardCore.Modules/OrchardCore.Media/SecureMediaPermissions.cs

```diff
@@ -0,0 +1,168 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Cache;
+using OrchardCore.Media.Services;
+using OrchardCore.Modules;
+using OrchardCore.Security.Permissions;
+
+namespace OrchardCore.Media
+{
+    [Feature("OrchardCore.Media.Security")]
+    public class SecureMediaPermissions : IPermissionProvider
+    {
+        // Note: The ManageMediaFolder permission grants all access, so viewing must be implied by it too.
+        public static readonly Permission ViewMedia = new("ViewMediaContent", "View media content in all folders", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewRootMedia = new("ViewRootMediaContent", "View media content in the root folder", new[] { ViewMedia });
+        public static readonly Permission ViewOthersMedia = new("ViewOthersMediaContent", "View others media content", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewOwnMedia = new("ViewOwnMediaContent", "View own media content", new[] { ViewOthersMedia });
+
+        private static readonly Permission _viewMediaTemplate = new("ViewMediaContent_{0}", "View media content in folder '{0}'", new[] { ViewMedia });
+
+        private static Dictionary<ValueTuple<string, string>, Permission> _permissionsByFolder = new();
+        private static readonly char[] _trimSecurePathChars = ['/', '\\', ' '];
+        public static readonly ReadOnlyDictionary<string, Permission> PermissionTemplates = new(new Dictionary<string, Permission>()
+        {
+            { ViewMedia.Name, _viewMediaTemplate },
+        });
```

Comment:
```suggestion
        // Note: The ManageMediaFolder permission grants all access, so viewing must be implied by it too.
        public static readonly Permission ViewMedia = new("ViewMediaContent", "View media content in all folders", new[] { Permissions.ManageMediaFolder });
        public static readonly Permission ViewRootMedia = new("ViewRootMediaContent", "View media content in the root folder", new[] { ViewMedia });
        public static readonly Permission ViewOthersMedia = new("ViewOthersMediaContent", "View others media content", new[] { Permissions.ManageMediaFolder });
        public static readonly Permission ViewOwnMedia = new("ViewOwnMediaContent", "View own media content", new[] { ViewOthersMedia });
        public static readonly ReadOnlyDictionary<string, Permission> PermissionTemplates = new(new Dictionary<string, Permission>()
        {
            { ViewMedia.Name, _viewMediaTemplate },
        });

        private static readonly Permission _viewMediaTemplate = new("ViewMediaContent_{0}", "View media content in folder '{0}'", new[] { ViewMedia });
        private static readonly char[] _trimSecurePathChars = ['/', '\\', ' '];

        private static Dictionary<ValueTuple<string, string>, Permission> _permissionsByFolder = new();
```

Reply:
Not possible due to initialization order...

Reply:
I will check